compiler/circle-mpqsolver/src/ErrorApproximator.cpp

```diff
@@ -0,0 +1,318 @@
+}
+
+bool getShape(CircleNode *circle_node, std::vector<uint32_t> &shape)
+{
+  if (circle_node->shape_status() == ShapeStatus::VALID)
+  {
+    auto rank = circle_node->rank();
+    if (rank != 4)
+      return false;
+
+    shape.resize(rank);
+    for (int i = 0; i < rank; i++)
+    {
+      shape[i] = circle_node->dim(i).value();
+    }
+    return true;
+  }
+
+  return false;
+}
+
+} // namespace
+
+/**
+ * How Approximate works?
+ *
+ * Currently it works just for convolution layers, but may be generalized for other types as well.
+ * See discussion at https://github.com/Samsung/ONE/pull/10170#discussion_r1042246598
+ * Convolution is just a matrix multiplication.
+ * While quantizing we introduce quantization error into convolution operand (activations) as well
+ * as into convolution weights. A_q * W_q = (A + q_err(A)) * (W + q_err(W)) = A * W + A * q_err(W) +
+ * W * q_err(A), q_err - quantization error, W - weight matrix, A - activations from previous layer,
+ * so quantization error of matrix multiplication is A * q_err(W) + W * q_err(A).
+ * Estimating its largest value we get
+ * ~ number_of_additions * (A_max * (W_max - W_min) / 255 + W_max * (A_max - A_min) / 255)
+ * The following code tries to get total error for quantizing convolution node into Q8.
+ * It's just an heuristic (Metric sensitivity depends highly on derivatives as well).
+ */
+float ErrorApproximator::approximate(CircleNode *node)
+{
+  auto circle_node = loco::must_cast<luci::CircleNode *>(node);
+  auto name = circle_node->name();
+  CircleConst *weights = nullptr;
+  uint32_t volume = 0;
+  std::vector<uint32_t> cir_shape;
+  if (getShape(circle_node, cir_shape) && cir_shape.size() == 4)
+  {
+    volume = cir_shape[1] * cir_shape[2] * cir_shape[0];
+  }
+
+  if (auto conv = dynamic_cast<luci::CircleConv2D *>(circle_node))
+  {
+    weights = loco::must_cast<luci::CircleConst *>(conv->filter());
+  }
+  else if (auto dconv = dynamic_cast<luci::CircleDepthwiseConv2D *>(circle_node))
+  {
+    weights = loco::must_cast<luci::CircleConst *>(dconv->filter());
+  }
+  else if (auto tconv = dynamic_cast<luci::CircleTransposeConv *>(circle_node))
+  {
+    weights = loco::must_cast<luci::CircleConst *>(tconv->filter());
+  }
+
+  if (weights != nullptr)
+  {
+    float ci_min = 0.f;
+    float ci_max = 0.f;
+
+    auto preds = loco::preds(node);
+    for (const auto &pred : preds)
+    {
+      auto parent_node = loco::must_cast<const luci::CircleNode *>(pred);
+      if (has_min_max(parent_node))
+      {
+        auto quantparam = parent_node->quantparam();
+        if (quantparam->min.size() > 0)
+        {
+          ci_min = quantparam->min[0];
+          ci_max = quantparam->max[0];
+        }
+      }
+    }
+
+    // channel-size min, max
+    std::vector<float> min_values;
+    std::vector<float> max_values;
+    {
+      int32_t channel_dim_index = 0;
+      cal_minmax_per_channel(weights, min_values, max_values, channel_dim_index);
```

Comment:
(optional)
1. Seems to me that `channel_dim_index` doesn't have to be a parameter of `cal_minmax_per_channel` method.
2. How about use convolution nodes having `opcode` instead of `weights`. Current codes have to check its origin(conv or dconv, etc) everytime.
3. `get_channel_dim_index` can be revisited. Simply getting only channel dim index.

With above consideration, codes would look like below.
```cxx
static uint32_t get_channel_dim_index(CircleNode *node)
{
  uint32_t index = 0;
  auto opcode = node->opcode();
  switch (opcode)
  {
    case luci::CircleOpcode::CONV_2D:
    case luci::CircleOpcode::TRANSPOSE_CONV:
    case luci::CircleOpcode::FULLY_CONNECTED:
    {
      index = 0;
    }
    case luci::CircleOpcode::DEPTHWISE_CONV_2D:
    {
      index = 3;
    }
    default:
      throw std::runtime_error("Failed to find channel index in " + node->name());
  }
  
  return index;
}

static bool set_weight_dim(CircleNode *node, loco::TensorShape dimension)
{
  auto opcode = node->opcode();
  switch (opcode)
  {
    case luci::CircleOpcode::CONV_2D:
    case luci::CircleOpcode::TRANSPOSE_CONV:
    case luci::CircleOpcode::DEPTHWISE_CONV_2D:
    {
      assert(node->rank() == 4);
      dimension.rank(node->rank());
      dimension.dim(0).set(node->dim(0).value());
      dimension.dim(1).set(node->dim(1).value());
      dimension.dim(2).set(node->dim(2).value());
      dimension.dim(3).set(node->dim(3).value());
    }
    case luci::CircleOpcode::FULLY_CONNECTED:
    {
      assert(node->rank() == 2);
      dimension.rank(4);
      dimension.dim(0).set(node->dim(0).value());
      dimension.dim(1).set(1); // Set FC layer like CONV
      dimension.dim(2).set(1);
      dimension.dim(3).set(node->dim(1).value());
    }
    default:
      return false;
  }

  return true;
}

static loco::Node *get_weight(CircleNode *node)
{
  loco::Node *weight = nullptr;
  auto opcode = node->opcode();
  switch (opcode)
  {
    case luci::CircleOpcode::CONV_2D:
    {
      auto conv = loco::must_cast<CircleConv2D *>(node);
      weight = conv->filter();
    }
    case luci::CircleOpcode::DEPTHWISE_CONV_2D:
    {
      auto dconv = loco::must_cast<CircleDepthwiseConv2D *>(node);
      weight = dconv->filter();
    }
    case luci::CircleOpcode::TRANSPOSE_CONV:
    {
      auto tconv = loco::must_cast<CircleTransposeConv *>(node);
      weight = tconv->filter();
    }
    case luci::CircleOpcode::FULLY_CONNECTED:
    {
      auto fc = loco::must_cast<CircleFullyConnected *>(node);
      weight = fc->weights();
    }
    default:
    {
      break;
    }
  }
  return weight;
}

static void iterate_per_channel(CircleNode *node, IterFunc func)
{
  loco::TensorShape dimension;
  set_weight_dim(node, dimension);
  uint32_t indices[4] = {
    0,
  };

  auto channel_dim_index = get_channel_dim_index(node);

  for (indices[0] = 0; indices[0] < dimension.dim(0).value(); indices[0]++)
  {
    for (indices[1] = 0; indices[1] < dimension.dim(1).value(); indices[1]++)
    {
      for (indices[2] = 0; indices[2] < dimension.dim(2).value(); indices[2]++)
      {
        for (indices[3] = 0; indices[3] < dimension.dim(3).value(); indices[3]++)
        {
          func(indices, dimension, channel_dim_index);
        }
      }
    }
  }
}

static void cal_minmax_per_channel(CircleNode *node, std::vector<float> &min,
                                   std::vector<float> &max)
{
  loco::TensorShape dimension;
  set_weight_dim(node, dimension);

  auto channel_dim_index = get_channel_dim_index(node);
  auto size = dimension.dim(channel_dim_index).value();

  std::vector<bool> has_min_max_value(size, false);
  min.resize(size);
  max.resize(size);

  CircleConst *weight = loco::must_cast<CircleConst *>(get_weight(node));
  auto cal_minmax = [&](uint32_t *indices, loco::TensorShape &dimension, int channel_dim_index) {
    int channel_idx = indices[channel_dim_index];
    auto data = weight->at<loco::DataType::FLOAT32>(cal_offset(dimension, indices));
    if (has_min_max_value[channel_idx])
    {
      min[channel_idx] = data < min[channel_idx] ? data : min[channel_idx];
      max[channel_idx] = data > max[channel_idx] ? data : max[channel_idx];
    }
    else
    {
      min[channel_idx] = data;
      max[channel_idx] = data;
      has_min_max_value[channel_idx] = true;
    }
  };

  iterate_per_channel(node, cal_minmax);
}
```

Reply:
I've reworked the whole `ErrorEpproximator.cpp` module according to @mhs4670go suggestions. I've discovered some bugs in computations. Seems like they are fixed for now. Still it correctly predicts that input nodes of `mobilenet_v1` should be quantized to Q16, and for `cartoon` model predicts the opposite (Q16 should be at output).